@abi checking

beccadax opened the pull request:
This PR adds logic to confirm that the ABI-only declaration in an `@abi` attribute is compatible with its counterpart—that is, that they are similar enough to one another that callees who expect the ABI-only decl will be able to successfully call the API decl. For instance, it will diagnose differences in the arity or generic signature, but also many more subtle differences.

It attempts to allow any differences that the checker knows are immaterial to call compatibility, such as:

* Different declaration names, argument labels, etc.
* Tuples with different labels but otherwise the same element type
* `Array<T>` instead of `T...` in parameters
* Uses of marker protocols or `sending`
* Parameter conventions that are mangled differently but represent the same memory management behavior

However, especially for type differences, these are effectively opt-in—there may be situations where an ABI-only decl is compatible with its API counterpart, but the compiler can't detect that and rejects it.

The compiler will also diagnose if there's anything unnecessary in the `@abi` declaration. This includes some very common things, such as default arguments, access control modifiers, and `@available` attributes. The compiler will offer fix-its to remove them.

**This PR doesn't actually adopt the feature yet. Standard library adoption has been cherry-picked atop this branch in #79937.**

-------

**Reviewers**: This is a large PR, but mostly owing to the fact that it specifies and tests behaviors for tons of attributes. I have attempted to split it into manageable commits, but this was done after the fact, so there's a little bit of wonkiness in the tests (`expected-error`s with diagnostics we haven't added yet, commits that make certain attributes irrelevant but don't actually ban them). It *should* all be tested in the end.

One issue I'm particularly worried about is whether I've forgotten any aspects of a declaration that need to be checked. Types will tend to fail safe, and there are mechanisms to make sure all attributes have a specified behavior, but it's totally possible for me to e.g. forget to check the failability of initializers (this actually happened, though I caught it myself) and thereby accidentally allow an ABI incompatibility to slip through.

beccadax committed:
[NFC] Fix ImplementsAttr::clone()

Correct an issue with `ImplementsAttr` that would come up if you ever tried to clone an attribute that had been deserialized.

No tests because there’s nothing in the compiler yet that might actually do so.

beccadax committed:
[NFC] Hoist getTypeSourceRangeForDiagnostics()

Allows code to get this for any AbstractStorageDecl.

beccadax committed:
[NFC] Expose default param specifier computation

Make this logic accessible outside of the mangler.

beccadax committed:
[NFC] Fix AllowFeatureSuppressionAttr atLoc

beccadax committed:
Support `@abi` on subscripts

And make sure we reject it on `deinit`s and accessors.

beccadax commented:
@swift-ci please test

beccadax committed:
Add flags to specify `@abi` behavior of attributes

This PR adds a set of DeclAttr.def flags for specifying how a given attribute interacts with `@abi`, and declares a behavior for each existing attribute. Future attributes will be *required* to declare an `@abi` behavior lest they fail a static assert.

Note that the behavior is not actually enforced in this commit—it is merely specified here.

beccadax committed:
[NFC] Extract Decl::getExplicitObjCName()

Create a helper method which looks for an `@objc` attribute with an explicit name and returns it, and adopt it in various existing places.

beccadax committed:
Inherit access control in `@abi`

ABI-only declarations now inherit access control modifiers like `public` or `private(set)`, as well as `@usableFromInline` and `@_spi`, from their API counterpart. This means these attributes and modifiers don’t need to be specified in an `@abi` attribute.

Very few tests because we aren’t yet enforcing the absence of these attributes.

beccadax committed:
Inherit `override` in `@abi`

ABI-only decls will now inherit `override` from their API counterpart; the keyword is unnecessary and `getOverriddenDecls()` will get the ABI counterparts of the decls the API counterpart overrides.

No tests because we don’t have the enforcement mechanism in yet.

beccadax committed:
Don’t add HasStorageAttr to ABI-only decls

It’s unnecessary, shouldn’t be serialized into module interfaces, and Swift doesn’t know how to compute it for an ABI-only decl since it doesn’t have accessors or an initial value.

No tests because enforcement isn’t in yet.

beccadax committed:
[NFC] Check that SILGen doesn’t use ABI-only decls

Adds assertions in various places where properties that can vary between ABI-only decls and their counterparts—particularly function and parameter attributes—are handled in SILGen, ensuring that we don’t accidentally end up processing ABI-only decls there.

beccadax committed:
Add DeclAttribute::isEquivalent()

Adds a new method to `DeclAttribute` which can compare two attributes and see if they would be “equivalent” in light of the given decl. “Equivalent” here means that they would have the same effect on the declaration; for instance, two attrs with different source locations can be equivalent, and two attributes with the same arguments in a different order are equivalent if the ordering of those argumetns is not semantically equivalent.

This capability is not yet used in this commit, but in a future commit `@abi` will check if certain attributes are equivalent or not.

beccadax commented:
@swift-ci please test

beccadax commented:
The test failures on https://github.com/swiftlang/swift/pull/79466/commits/96739a94277a4f867ee0e20244ac335516c53ee5 will require minor tweaks (mostly to tests), but this is basically ready to review.

beccadax commented:
@swift-ci please test

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -8570,3 +8501,1036 @@ ArrayRef<VarDecl *> InitAccessorReferencedVariablesRequest::evaluate(
+  }
+
+  static GenericSignature normalizeGenericSignature(GenericSignature original) {
+    // FIXME: Are there other ABI-tolerable generic signature differences?
```
You can actually generalize this considerably, if you feel like it. The "ABI" of a generic signature is basically a list of calling convention parameters, for the metadata and witness tables.

There is a metadata argument for each reduced generic parameter, and a witness table for every conformance requirement.

You want to build a new generic signature from the same list of generic parameters, but the following requirements:
- For each generic parameter in the original signature such that `!sig->isReducedType(param)`, add a requirement `param == ()` to the new signature (any other concrete type on the RHS will also do).
- For each conformance requirement to a non-@objc, non-marker protocol in the original signature, add the same conformance requirement to the new signature.
- Ignore all other requirements.

beccadax committed:
Diagnose mismatched typed throws in `@abi`

beccadax committed:
Improve diagnosis of generic types in `@abi`

A same-type constraint in an enclosing `where` clause will eliminate a generic parameter’s ABI impact. Teach `ABIDeclChecker::checkType()` about this so it can handle a known-unsupported case for `Swift.Result.init(catching:)`.

beccadax commented:
@swift-ci please test

beccadax committed:
Filter bad attrs out of module interface `@abi`s

When printing an `@abi` attribute’s decl, we now filter out any attrs that are not valid in that position. Fixes a broken test.

beccadax committed:
Check attributes in `@abi` attr

This commit compares the attributes on the decl inside the `@abi` attribute to those in the decl it’s attached to, diagnosing ABI-incompatible differences. It also rejects many attributes that don’t need to be specified in the `@abi` attribute, such as ObjC-ness, access control, or ABI-neutral traits like `@discardableResult`, so developers know to remove them.

beccadax committed:
Inherit availability in `@abi`

ABI-only declarations now inherit `@available`, `@backDeployed`, etc. from their ABI counterpart. This will make it unnecessary to specify these attributes in `@abi`. Also some changes to make sure we suggest inserting `@available` in the correct place.

No tests because the enforcement is not yet in.

beccadax committed:
Inherit @objc, dynamic, @implementation in `@abi`

ABI-only declarations now query their API counterpart for things like `isObjC()`, their ObjC name, dynamic status, etc. This means that `@objc` and friends can simply be omitted from an `@abi` attribute.

No tests in this commit since attribute checking hasn’t landed yet.

beccadax committed:
Make `@abi` type diagnostics more specific

Specify whether the type with the problem is a result type, parameter type (and which parameter), etc.

beccadax committed:
Type check `@abi` decls (sans attrs)

This commit compares the decl inside the `@abi` attribute to the decl it’s attached to, diagnosing ABI-incompatible differences. It does not yet cover attributes, which are a large undertaking.

beccadax commented:
In addition to rebasing, this force-push extracts most of the `@abi` checking into a new source file.

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci smoke test

beccadax commented:
@swift-ci smoke test

beccadax commented:
@swift-ci smoke test

beccadax committed:
Tweak `@abi` `@lifetime` test

Remove a diagnostic that isn’t occurring anymore.

beccadax commented:
@swift-ci smoke test

nkcsgexi commented:
@swift-ci smoke test

beccadax commented:
Merge conflict wasn't quite as easy to resolve as it looked.

@swift-ci smoke test

beccadax commented:
@swift-ci please smoke test Linux platform

nkcsgexi commented:
Seems we have hit:
```
/home/build-user/swift/include/swift/AST/DeclAttr.def:876:1: error: static assertion failed due to requirement 'DeclAttribute::hasOneBehaviorForConstInitialized(DeclAttribute::InABIAttrMask)': constInitialized needs to specify exactly one of ForbiddenInABIAttr, UnconstrainedInABIAttr, EquivalentInABIAttr, InferredInABIAttr, or UnreachableInABIAttr
```

beccadax committed:
Add `@abi` behavior for `@constInitialized`

beccadax commented:
@swift-ci please smoke test

beccadax commented:
@swift-ci please smoke test

nkcsgexi commented:
ah, we've hit another conflict:
```
DECL_ATTR(execution, Execution,
<<<<<<< abi-inspected-your-appearance
  OnAbstractFunction,
  ABIBreakingToAdd | ABIBreakingToRemove | APIBreakingToAdd | APIBreakingToRemove | UnconstrainedInABIAttr,
=======
  OnFunc | OnConstructor | OnSubscript | OnVar,
  ABIBreakingToAdd | ABIBreakingToRemove | APIBreakingToAdd | APIBreakingToRemove,
>>>>>>> main
  166)
```

beccadax commented:
@swift-ci please smoke test

nkcsgexi commented:
🎉

drodriguez commented on test/attr/attr_abi.swift:
```diff
@@ -1,7 +1,21 @@
+// REQUIRES: swift_feature_StrictMemorySafety
+// REQUIRES: swift_feature_SymbolLinkageMarkers
+
+import _Differentiation
```
Missing a `REQUIRES: differentiable_programming` if one wants to import `_Differentiation`.

drodriguez commented on test/attr/attr_abi.swift:
```diff
@@ -1,7 +1,21 @@
+
+import _Differentiation
+
+import Distributed
```
Missing a `REQUIRES: distributed` for this `import Distributed` as well.
